pkg/conn/scheduler_utils.go

```diff
@@ -0,0 +1,169 @@
+
+import (
+	"context"
+	"math"
+
+	"github.com/pingcap/errors"
+
+	"github.com/pingcap/br/pkg/utils"
+)
+
+// clusterConfig represents a set of scheduler whose config have been modified
+// along with their original config.
+type clusterConfig struct {
+	// Enable PD schedulers before restore
+	scheduler []string
+	// Original scheudle configuration
+	scheduleCfg map[string]interface{}
+}
+
+var (
+	schedulers = map[string]struct{}{
+		"balance-leader-scheduler":     {},
+		"balance-hot-region-scheduler": {},
+		"balance-region-scheduler":     {},
+
+		"shuffle-leader-scheduler":     {},
+		"shuffle-region-scheduler":     {},
+		"shuffle-hot-region-scheduler": {},
+	}
+	pdRegionMergeCfg = []string{
+		"max-merge-region-keys",
+		"max-merge-region-size",
+	}
+	pdScheduleLimitCfg = []string{
+		"leader-schedule-limit",
+		"region-schedule-limit",
+		"max-snapshot-count",
+	}
+)
+
+func addPDLeaderScheduler(ctx context.Context, mgr *Mgr, removedSchedulers []string) error {
+	for _, scheduler := range removedSchedulers {
+		err := mgr.AddScheduler(ctx, scheduler)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func restoreSchedulers(ctx context.Context, mgr *Mgr, clusterCfg clusterConfig) error {
+	if err := addPDLeaderScheduler(ctx, mgr, clusterCfg.scheduler); err != nil {
+		return errors.Annotate(err, "fail to add PD schedulers")
+	}
+	mergeCfg := make(map[string]interface{})
+	for _, cfgKey := range pdRegionMergeCfg {
+		value := clusterCfg.scheduleCfg[cfgKey]
+		if value == nil {
+			// Ignore non-exist config.
+			continue
+		}
+		mergeCfg[cfgKey] = value
+	}
+	if err := mgr.UpdatePDScheduleConfig(ctx, mergeCfg); err != nil {
+		return errors.Annotate(err, "fail to update PD merge config")
+	}
+
+	scheduleLimitCfg := make(map[string]interface{})
+	for _, cfgKey := range pdScheduleLimitCfg {
+		value := clusterCfg.scheduleCfg[cfgKey]
+		if value == nil {
+			// Ignore non-exist config.
+			continue
+		}
+		scheduleLimitCfg[cfgKey] = value
+	}
+	if err := mgr.UpdatePDScheduleConfig(ctx, scheduleLimitCfg); err != nil {
+		return errors.Annotate(err, "fail to update PD schedule config")
+	}
+	return nil
+}
+
+// RemoveSchedulers removes the schedulers that may slow down BR speed.
+// TODO make each step returns a function that can restore schedulers it has removed.
+func (mgr *Mgr) RemoveSchedulers(ctx context.Context) (utils.UndoFunc, error) {
+	// Remove default PD scheduler that may affect restore process.
+	existSchedulers, err := mgr.ListSchedulers(ctx)
+	if err != nil {
+		return utils.Nop, nil
+	}
+	needRemoveSchedulers := make([]string, 0, len(existSchedulers))
+	for _, s := range existSchedulers {
+		if _, ok := schedulers[s]; ok {
+			needRemoveSchedulers = append(needRemoveSchedulers, s)
+		}
+	}
+	scheduler, err := removePDLeaderScheduler(ctx, mgr, needRemoveSchedulers)
+	if err != nil {
+		return utils.Nop, nil
+	}
+
+	stores, err := mgr.GetPDClient().GetAllStores(ctx)
+	if err != nil {
+		return utils.Nop, err
```

Comment:
i think you shouldn't return Nop given the removePDLeaderScheduler has succeeded.